### PR TITLE
[WIP] Persist DAG on disk

### PR DESCRIPTION
**Pull request by itegulov**
## Overview
Implements the rest parts of `BlockDagStorage`: `childMap` storage, `dataLookup` storage and `toposort` storage.

### Which JIRA issue does this PR relate to? If there is not a JIRA issue addressing this work, please create one now and add the link here.
https://rchain.atlassian.net/browse/RHOL-539

### Complete this checklist before you submit the PR
- [ ] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [x] If this PR adds a new feature, this PR includes tests related to this feature.
- [ ] You assigned one person to review this PR

### If you are not a member of the core development team, please confirm:
- [ ] You signed the commit. Merging requires a signature. Please see the [developer wiki](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain) for instructions.
- [ ] Your GitHub account is also an account with [Travis CI](https://travis-ci.org). Unit tests will not run on your PR unless you have an account with Travis. Merge requires passed unit tests.

### Notes
Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else.

### CODE REVIEW

File: block-storage/src/main/scala/coop/rchain/blockstorage/BlockMetadata.scala [hunk 1]
```diff
@@ -1,19 +1,54 @@
...
     parents: List[BlockHash],
     sender: ByteString,
     justifications: List[Justification],
-    weightMap: Map[ByteString, Long],
+    weightMap: Map[Validator, Long],
     blockNum: Long,
     seqNum: Int
-)
+) {
+  def toByteString: ByteString = {
```
Comment (birchmd) [inline block-storage/src/main/scala/coop/rchain/blockstorage/BlockMetadata.scala]: I wonder if this method, along with `fromBytes` would be an argument for just making this class be part of the protobuf model (then we would get them for free).

**Comment by itegulov** [comment]
I am almost finished with this: implemented checkpoint loading, storage for block metadata (`childMap` and `topoSort` are not stored separately and are instead built from block metadata log) and addressed @birchmd's comments about effectful operations. I will add more corner-case tests and implement `checkpoint()` function properly shortly.

@birchmd PTAL

File: block-storage/src/main/scala/coop/rchain/blockstorage/BlockMetadata.scala (hunk 1 above)
Comment (itegulov) [reply block-storage/src/main/scala/coop/rchain/blockstorage/BlockMetadata.scala]: I think this makes sense. It will also allow us to modify `BlockMetadata` structure seamlessly in the future.

File: block-storage/src/test/scala/coop/rchain/blockstorage/BlockDagStorageTest.scala [hunk 2]
```diff
@@ -43,116 +70,310 @@ class BlockDagFileStorageTest extends BlockDagStorageTest {
...
+              .map(_.blockHash)
+              .toSet
+          )
+        contains shouldBe true
+    }
+    latestMessageHashes shouldBe blockElements.map(b => b.sender -> b.blockHash).toMap
+    latestMessages shouldBe blockElements.map(b => b.sender      -> BlockMetadata.fromBlock(b)).toMap
+
+    def normalize(topoSort: Vector[Vector[BlockHash]]): Vector[Vector[BlockHash]] =
+      if (topoSort.size == 1 && topoSort.head.isEmpty)
```
Comment (Primetalk) [inline block-storage/src/test/scala/coop/rchain/blockstorage/BlockDagStorageTest.scala]: `topoSort.filter(_.nonEmpty)`?

File: block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagStorage.scala [hunk 3]
```diff
@@ -22,7 +22,7 @@ trait BlockDagRepresentation[F[_]] {
   def lookup(blockHash: BlockHash): F[Option[BlockMetadata]]
   def contains(blockHash: BlockHash): F[Boolean]
   def topoSort(startBlockNumber: Long): F[Vector[Vector[BlockHash]]]
-  def topoSortTail(tailLength: Long): F[Vector[Vector[BlockHash]]]
+  def topoSortTail(tailLength: Int): F[Vector[Vector[BlockHash]]]
```
[the remark this answers is not in the document]
Comment (itegulov) [reply block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagStorage.scala]: `tailLength` basically represents the length of the resulting vector. Since we can't create vectors longer than 2^31-1, it does not make sense to allow to specify a `Long` length for it. On the other hand, `startBlockNumber` just represents the offset which could potentially be larger than 2^31-1, but it is okay since the entire data structure won't be stored in memory.

File: block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala [hunk 4]
```diff
@@ -382,16 +374,12 @@ final class BlockDagFileStorage[F[_]: Concurrent: Sync: Log: BlockStore] private
...
-            latestMessagesLogOutputStream.close()
-            latestMessagesLogOutputStream
-          }
-      _ <- blockMetadataLogOutputStreamRef.update { blockMetadataLogOutputStream =>
-            blockMetadataLogOutputStream.close()
-            blockMetadataLogOutputStream
-          }
-      _ <- lock.release
+      _                             <- lock.acquire
+      latestMessagesLogOutputStream <- latestMessagesLogOutputStreamRef.get
```
Comment (Primetalk) [inline block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala]: INFO: 
```scala
_ <- latestMessagesLogOutputStreamRef.get.map(_.close())
```

**Comment by itegulov** [comment]
@Primetalk thanks for your review, it was very helpful!
